Implement react and update_message for EmbeddedBotHandler

PIG208 opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
#17860

**Testing plan:** <!-- How have you tested? -->
Unit tests were added for testing `react` and `update_message`.

**Side notes**
The implementation for `update_message` is largely the same as that for the equivalent view `update_message_backend` as I extract part of it to reuse the code.

This PR doesn't yet introduce custom exceptions to embedded bot handlers, but it's worth considering to add it in the future as Anders [suggests](https://chat.zulip.org/#narrow/stream/49-development-help/topic/Error.20Handling.20for.20EmbeddedBotHandler/near/1150009).
Currently, the implementation just returns a dictionary object containing the error message to be consistent with `ExternalBotHandler`.

<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

PIG208 commented:
Don't know why `diff` is somehow corrupted on `message_edit.py`.

timabbott commented:
I merged the first commit, thanks for catching and tracing the history!  For the `message_edit` commit, I think the issue is that you reordered functions as part of extracting that function -- I'd encourage you to split that to two commits, first one that just reorders code (if you think the reordering was a good idea, otherwise just undo it); `git show` can color-highlight moved code, making it really easy to review a commit that's supposed to just move code without changes.  And then your actual changes will be much clearer.

PIG208 commented:
OK, I think I have fixed that by breaking the commit into two and move it below the original function instead. It appears that `diff` prefers to move the header instead of the body when the body is too long.

LoopThrough-i-j commented on zerver/views/message_edit.py:
```diff
@@ -113,6 +113,48 @@ def update_message_backend(
+    return json_success()
+
+
+def check_update_message(
```
I think this should be in `zerver/lib/actions.py`

LoopThrough-i-j replied:
Though I don't see a need for the new function.

PIG208 replied:
Because I need to reuse the check and update logic for embedded bots. But yes, I think we can move this to `zerver/lib/actions.py`.

LoopThrough-i-j commented on zerver/views/message_edit.py:
```diff
@@ -114,10 +114,10 @@ def update_message_backend(
 ) -> HttpResponse:
     if not user_profile.realm.allow_message_editing:
-        return json_error(_("Your organization has turned off message editing"))
+        raise JsonableError(_("Your organization has turned off message editing"))
```
I don't think we would like to throw errors in API calls.

PIG208 replied:
Then would `return json_error` be a better alternative here? I found that other functions in `zerver/lib/actions.py` do raise exceptions, perhaps it's reasonable to do so in this context.

LoopThrough-i-j replied:
Yeah, that's true in actions we throw errors but the api client never faces an error, we would always want the api clients to get a proper response for the error thrown.

LoopThrough-i-j commented on zerver/lib/bot_lib.py:
```diff
@@ -83,7 +91,26 @@ def identity(self) -> BotIdentity:
+                reaction_type,
+            )
+        except JsonableError as e:
+            return generate_error_msg(e.msg)
```
I don't think `EmbeddedBotHandler` should handle exceptions after the latest discussions.

PIG208 replied:
OK, then I guess we will return an empty dictionary instead.

LoopThrough-i-j replied:
It is the `worker` that should do that if required.

LoopThrough-i-j commented on zerver/lib/bot_lib.py:
```diff
@@ -137,8 +164,31 @@ def send_reply(
+                content=message.get("content", None),
+            )
+        except JsonableError as e:
+            return generate_error_msg(e.msg)
```
I don't think `EmbeddedBotHandler` should handle exceptions here as well.

LoopThrough-i-j commented:
Thanks for working on this. I have mentioned some points for discussion.
Moreover, I would like you to have 2 separate commits for the `EmbeddedBotHandler`. one for `message_update` and the other for reactions, and each commit should have the tests for the feature implemented.

PIG208 commented:
Thank you for reviewing! I think another reasonable change here might be moving `check_update_message` to `zerver/lib/action.py`.

PIG208 commented:
Since the last review, I found that there was some problem with `update_message` as it wrongly considers `id` to be the key for the message's id. I also fixed the test case for `update_message` so it now correctly checks that the message is updated instead of being sent again. 
Apart from that, previously, `send_message` and `reply_message` return `None`, which will make the incrementor fail to find the message to update. So I changed that as well to return the id as our API does.

LoopThrough-i-j reviewed (commented):
I think we would like you to rebase instead of merging the master

LoopThrough-i-j commented on zerver/lib/bot_lib.py:
```diff
@@ -85,19 +85,19 @@ def identity(self) -> BotIdentity:
         return {}  # Not implemented
 
-    def send_message(self, message: Dict[str, Any]) -> None:
+    def send_message(self, message: Dict[str, Any]) -> Dict[str, Any]:
```
Not sure if we need to return a `dict`, why not just `int`

PIG208 replied:
I do that because the protocol is defined as `def send_message(self, message: Dict[str, Any]) -> Optional[Dict[str, Any]]:`

LoopThrough-i-j replied:
Ok we will need to make proper changes soon.

PIG208 commented:
I think everything should be working now. The PR is ready for another review

LoopThrough-i-j commented on zerver/tests/test_embedded_bot_system.py:
```diff
@@ -52,6 +53,31 @@ def test_stream_message_to_embedded_bot(self) -> None:
         self.assertEqual(display_recipient, "Denmark")
 
+    def test_embedded_bot_react_to_stream_message(self) -> None:
+        assert self.bot_profile is not None
```
A test for failing reaction will help, just to see we get the appropriate Exception raised.

PIG208 replied:
Agree, I have added new test cases to support that.

LoopThrough-i-j commented on zerver/tests/test_embedded_bot_system.py:
```diff
@@ -52,6 +53,31 @@ def test_stream_message_to_embedded_bot(self) -> None:
+            ).exists()
+        )
+
+    def test_embedded_bot_react_to_pm(self) -> None:
```
A test for Failing reaction will help here too.

PIG208 replied:
Actually, this shouldn't matter whether it's pm or stream message, as the exceptions that we are interested in occur before the reaction is actually added in `do_add_reaction`.

LoopThrough-i-j commented on zerver/lib/bot_lib.py:
```diff
@@ -83,7 +89,22 @@ def identity(self) -> BotIdentity:
     def react(self, message: Dict[str, Any], emoji_name: str) -> Dict[str, Any]:
-        return {}  # Not implemented
+        # Currently, instead of raising exceptions, we return dictionaries containing the error
+        # message for bot handlers. But in the future, these should be replaced by python exceptions.
```
I think now we do raise exceptions.

PIG208 replied:
Yes, I have updated that in the new PR.

LoopThrough-i-j commented on zerver/tests/test_embedded_bot_system.py:
```diff
@@ -78,6 +86,21 @@ def test_embedded_bot_react_to_pm(self) -> None:
             ).exists()
         )
 
+    def test_embedded_bot_update_message(self) -> None:
```
A test for exception would help.

PIG208 replied:
Added tests for that.

LoopThrough-i-j commented on zerver/lib/actions.py:
```diff
@@ -2567,6 +2568,151 @@ def check_schedule_message(
     return do_schedule_messages([send_request])[0]
 
 
+def check_update_message(
```
I think we would like to have some tests for this function, as it does a lot of valuable checks which we might want to be there for all forms of message edits.

PIG208 replied:
The `check_update_message` has been tested thoroughly in `test_message_edit`. There is no need to create new tests as the PR only extracts it without making significant changes in features.

timabbott replied:
Agreed.

LoopThrough-i-j commented:
Thanks for all the changes. Added a few more comments. I think we would like to have a few more tests as mentioned and that's all.

PIG208 commented:
@LoopThrough-i-j Thank you for spending so much time on reviewing this PR. I think it has become more complete now.

LoopThrough-i-j commented:
Thanks, things look better now. Great work soo far.

PIG208 commented:
@zulipbot add "bots"

PIG208 commented:
@zulipbot add "area: bots"

timabbott commented on zerver/lib/bot_lib.py:
```diff
@@ -69,6 +71,10 @@ class EmbeddedBotEmptyRecipientsList(Exception):
 
 
+class EmbeddedBotKeyNotFound(Exception):
+    pass
```
This can be a followup PR, but I think structurally we probably want these to be all subclasses of some exception with common exception handling logic that does something reasonable.

(I believe all of these would represent bugs in the embedded bot, so should log an error for server administrators to investigate)

PIG208 replied:
Yes. We can do `EmbeddedBotError` or something here. I think we should raise exceptions in a similar manner for external bots as well.

PIG208 replied:
I think we can create a protocol for any bot errors that can be applied to both embedded bots and external bots. For example, when the user is writing a bot, they will expect `update_message` to throw an error instead of returning a dictionary that looks like:
```
{
    "code": "BAD_REQUEST",
    "msg": "You don't have permission to edit this message",
    "result": "error"
}
```
Instead, we may have something that `BotError` that applies to both external bots and embedded bots. This will allow better error handling practice.

eeshangarg replied:
Since there is an expectation that a dict should be returned here, should we return a dict in the cases where an exception is raised at all?

PIG208 replied:
That's quite a good question, but we can't do both of them at the same time. I think the ideal case for us might be migrating to use these exceptions before making the error handling logic final. As after all, we would like to replace the dictionaries by exceptions (this part of the work might need to be completed on the python-zulip-api part.

PIG208 replied:
But speaking of compatibility, perhaps we can postpone adding the exceptions for now?

eeshangarg replied:
Hmm, maybe, I am honestly not sure! My initial sense is that just raising an exception without letting the user know anything kinda breaks expectations a little. But yes, we can probably postpone the work on exceptions and do it in a follow-up PR.

timabbott commented on zerver/lib/bot_lib.py:
```diff
@@ -83,7 +89,20 @@ def identity(self) -> BotIdentity:
+            reaction_type,
+        )
+
+        return {}
```
Is the right interface for this to return the details of the created reaction in the dictionary?  I can imagine a client wanting those to e.g. unreact naturally.

(Can be left for future work, since that's a coordinated change; we probably also want a `remove_reaction` thing eventually -- just thinking through the direction here).

PIG208 replied:
It should look like this,
```json
{
    "msg": "",
    "result": "success"
}
```
I will fix that.

PIG208 replied:
Yeah, `remove_reaction` can be added to our protocol in python-zulip-api.

PIG208 replied:
I have created an issue for this https://github.com/zulip/python-zulip-api/issues/684 in the Zulip API repo.

timabbott commented on zerver/tests/test_embedded_bot_system.py:
```diff
@@ -57,6 +60,42 @@ def test_stream_message_not_to_embedded_bot(self) -> None:
+            Reaction.objects.filter(
+                message=user_message,
+            ).exists()
+        )
```
Where is the test setup logic that configures the bot in question to actually call `react()`?

PIG208 replied:
The bot is a helloworld bot, which replies to the message and adds a reaction to it. I will note that in the comment.

timabbott commented:
@PIG208 can your rebase this?

timabbott commented:
I merged a couple prep commits as the series ending with 35fae794221ee658d904e0a9ef1dcf58d7901a9f.

PIG208 commented:
Yes. I will work on this once I'm available.

PIG208 commented:
The PR has been rebased with some minor changes including adding comments explaining the test setup for `react`, and changing the return value of `react` and `update_message` to match with our API documentation ([like this one](https://zulip.com/api/add-reaction)).

timabbott commented:
I merged the first commit as 7150fe5dc5bedac5c337ea542264df9398cbf637.  I think the react commit at least requires some comments/explanation as to why it's correct to call `do_add_reaction` rather than `check_add_reaction`.

timabbott commented on zerver/lib/bot_lib.py:
```diff
@@ -160,8 +161,26 @@ def send_reply(
+            send_notification_to_old_thread=message.get("send_notification_to_old_thread", False),
+            send_notification_to_new_thread=message.get("send_notification_to_new_thread", False),
+            content=message.get("content", None),
+        )
```
I'd like us to do proper parsing here using `check_dict(...)` to make sure that we're not passing unexpected types for these parameters into `check_update_message`.

PIG208 commented:
> I merged the first commit as [7150fe5](https://github.com/zulip/zulip/commit/7150fe5dc5bedac5c337ea542264df9398cbf637). I think the react commit at least requires some comments/explanation as to why it's correct to call `do_add_reaction` rather than `check_add_reaction`.

`check_add_reaction` wasn't there when the commit is written and I think it should be reasonable to use it here. But I came across another problem: when calling `check_add_reactions`, `access_message` is called using the message id and the bot's user profile, but it appears that the bot user doesn't have access to private messages that are sent to them, which leads to an "invalid message(s)" error.

PIG208 commented:
I have added `check_dict` for both `react` and `update_message`. However, to have `check_add_reaction` available for embedded bots, we will first need to get #18117 merged.

eeshangarg reviewed (commented):
Hey @PIG208! Thanks for working on this! I feel like the other reviewers already caught a lot of what I would have said, so I didn't have much to add here.

A minor comment on the commit message for the second commit, you do not need to mention "with test cases added". In the Zulip repository, it is well known that a commit that makes a change will be accompanied by proper tests, so there is no need to mention that in the commit message.

Thanks again! Have a good day! :)

PIG208 commented on zerver/lib/bot_lib.py:
```diff
@@ -142,8 +168,37 @@ def send_reply(
+            send_notification_to_new_thread=message.get("send_notification_to_new_thread", False),
+            content=message.get("content", None),
+        )
+        return {"msg": "", "result": "success"}
```
Yes, we have to make sure that the implementation matches the outgoing webhook and the API documentation for compatibiity.
